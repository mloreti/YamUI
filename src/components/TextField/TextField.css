/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */

.y-text-field.ms-TextField--underlined .ms-TextField-fieldGroup {
  height: 2.5rem
}

.y-text-field.ms-TextField--underlined .ms-TextField-field {
  padding-left: 0px;
}

.y-text-field.ms-TextField--underlined.is-active .ms-TextField-wrapper {
  border-color: $borderColor__focused;
}

.y-text-field--with-prefix.ms-TextField--underlined .ms-TextField-field,
.y-text-field--with-prefix .ms-TextField-field {
  padding-left: .4rem;
}

.y-text-field--with-suffix.ms-TextField--underlined .ms-TextField-field,
.y-text-field--with-suffix .ms-TextField-field {
  padding-right: .4rem;
}

.y-text-field .ms-TextField-prefix,
.y-text-field .ms-TextField-suffix {
  padding: 0 .8rem;
}
