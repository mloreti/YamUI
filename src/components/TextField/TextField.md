### Examples

Basic
```js { "props": { "data-description": "basic" } }
<TextField
  label="First Name"
  value="Textfield filled"
  description="Small hint goes here"
/>
```

Disabled
```js { "props": { "data-description": "disabled" } }
<TextField
  value="Disabled"
  disabled
/>
```

With Placeholder
```js { "props": { "data-description": "with placeholder" } }
<TextField
  placeHolder="Textfield Placeholder..."
/>
```

With Error Message
```js { "props": { "data-description": "with error" } }
<TextField
  value="Error"
  errorMessage="Error hint goes here"
/>
```

With Error Message and Description.  (NOTE: The description is hidden.)
```js { "props": { "data-description": "with error and description" } }
<TextField
  value="Error"
  errorMessage="Error hint goes here"
  description="This should not be shown"
/>
```

With Prefix
```js { "props": { "data-description": "with prefix" } }
<TextField
  placeHolder="your-website.com"
  prefix="http://"
/>
```

With Suffix
```js { "props": { "data-description": "with suffix" } }
<TextField
  placeHolder="your e-mail"
  suffix="@contoso.com"
/>
```

Long Prefix
```js { "props": { "data-description": "long prefix" } }
<TextField
  placeHolder="placeholder text"
  prefix="long prefix long prefix long prefix long prefix long prefix long prefix long prefix"
/>
```

Long Prefix and Suffix
```js { "props": { "data-description": "long prefix and suffix" } }
<TextField
  placeHolder="placeholder text"
  prefix="long prefix long prefix long prefix long prefix long prefix long prefix long prefix"
  suffix="long suffix long suffix long suffix long suffix long suffix long suffix long suffix"
/>
```

Short Prefix and Long Suffix
```js { "props": { "data-description": "short prefix and long suffix" } }
<TextField
  placeHolder="placeholder text"
  prefix="short prefix"
  suffix="long suffix long suffix long suffix long suffix long suffix long suffix long suffix"
/>
```

Long Prefix and Short Suffix
```js { "props": { "data-description": "long prefix and short suffix" } }
<TextField
  placeHolder="placeholder text"
  prefix="long prefix long prefix long prefix long prefix long prefix long prefix long prefix"
  suffix="short suffix"
/>
```

Required:
```js { "props": { "data-description": "required" } }
<TextField
  placeHolder="Required Field"
  required
/>
```

Required with Label:
```js { "props": { "data-description": "required with label" } }
<TextField
  label={"Required Field"}
  required
/>
```

onChange:
```js { "props": { "data-description": "onChange" } }
<TextField
  onChange={action('change from `onChange` example')}
/>
```

Controlled onChange with debounce:
```js { "props": { "data-description": "controlled onChange with debounce" } }
class ControlledTextField extends React.Component {
  constructor() {
    super();
    this.state = {};
    this.handleChange = this.handleChange.bind(this);
  }

  render() {
    return (
      <TextField
        value={this.state.value}
        debouncedOnChange={(value) => action('textfield changed (debounced)')(value)}
        debouncedOnChangeTime={700}
        onChange={this.handleChange}
      />
    );
  }

  handleChange(value) {
    action('textfield changed')(value);
    this.setState({ value });
  }
}

<ControlledTextField />
```

Beside Button
```js { "props": { "data-description": "beside button" } }
<FixedGridRow>
  <FixedGridColumn><TextField/></FixedGridColumn>
  <FixedGridColumn fixed={true}><Button text="Submit" /></FixedGridColumn>
</FixedGridRow>
```

Underlined Variation
```js { "props": { "data-description": "basic underlined" } }
<TextField
  placeHolder="Textfield Placeholder..."
  underlined
/>
```

Underlined Variation With Error
```js { "props": { "data-description": "underlined with error" } }
<TextField
  errorMessage="Error hint goes here"
  placeHolder="Textfield Error"
  underlined
/>
```
