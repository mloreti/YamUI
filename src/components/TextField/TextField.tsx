/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import '../../yamui';
import * as React from 'react';
import { getBaseTextFieldProps, BaseTextFieldProps, ITextField } from '../BaseTextField';
import DebouncedOnChange, { DebouncedOnChangeProps, DebouncedOnChangePrivateProps } from '../../util/DebouncedOnChange';
import { TextField as FabricTextField } from 'office-ui-fabric-react/lib/TextField';
import '../BaseTextField/BaseTextField.css';
import './TextField.css';

export { ITextField };

export interface TextFieldProps extends BaseTextFieldProps, DebouncedOnChangeProps {
  /**
   * String for prefix.
   */
  prefix?: string;

  /**
   * String for suffix.
   */
  suffix?: string;

  /**
   * Whether or not the textfield is underlined.
   */
  underlined?: boolean;
}

/**
 * The TextField component enables a user to type text into an app. It's used to capture
 * a single line of text. The text displays on the screen in a simple, uniform format.
 */
class TextField extends React.Component<TextFieldProps & DebouncedOnChangePrivateProps> {
  public render() {
    return (
      <FabricTextField
        className={this.getClasses()}
        prefix={this.props.prefix}
        suffix={this.props.suffix}
        underlined={this.props.underlined}
        onChanged={this.props.unifiedOnChange}
        {...getBaseTextFieldProps(this.props)}
        componentRef={this.props.componentRef}
        maxLength={this.props.maxLength}
      />
    );
  }

  private getClasses() {
    const { className, prefix, suffix } = this.props;
    const classes = ['y-base-text-field', 'y-text-field'];

    if (className) {
      classes.push(className);
    }

    if (prefix) {
      classes.push('y-text-field--with-prefix');
    }

    if (suffix) {
      classes.push('y-text-field--with-suffix');
    }

    return classes.join(' ');
  }
}

export default (props: TextFieldProps) => {
  return <DebouncedOnChange {...props} component={TextField} />;
};
