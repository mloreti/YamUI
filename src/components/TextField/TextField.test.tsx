/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import * as React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import TextField, { TextFieldProps } from '.';

describe('<TextField />', () => {
  let component: ShallowWrapper<TextFieldProps>;

  describe('with default options', () => {
    beforeEach(() => {
      component = shallow(<TextField />)
        .dive()
        .dive();
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('with all options', () => {
    beforeEach(() => {
      component = shallow(
        <TextField
          prefix="PREFIX"
          suffix="SUFFIX"
          maxLength={100}
          underlined={true}
          className="CLASS"
          value="VALUE"
          label="LABEL"
          description="DESCRIPTION"
          disabled={true}
          required={true}
          errorMessage="ERROR MESSAGE"
          placeHolder="PLACEHOLDER"
          onFocus={jest.fn().mockName('onFocus')}
          onBlur={jest.fn().mockName('onBlur')}
          onMouseEnter={jest.fn().mockName('onMouseEnter')}
          onMouseLeave={jest.fn().mockName('onMouseLeave')}
        />,
      )
        .dive()
        .dive();
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });
});
