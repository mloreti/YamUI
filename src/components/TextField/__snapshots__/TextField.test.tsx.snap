// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> with all options matches its snapshot 1`] = `
<TextField
  aria-required={true}
  autoAdjustHeight={false}
  borderless={false}
  className="y-base-text-field y-text-field CLASS y-text-field--with-prefix y-text-field--with-suffix"
  deferredValidationTime={200}
  disabled={true}
  errorMessage="ERROR MESSAGE"
  label="LABEL"
  maxLength={100}
  multiline={false}
  onBeforeChange={[Function]}
  onBlur={[MockFunction onBlur]}
  onChanged={[Function]}
  onFocus={[MockFunction onFocus]}
  onGetErrorMessage={[Function]}
  onMouseEnter={[MockFunction onMouseEnter]}
  onMouseLeave={[MockFunction onMouseLeave]}
  onNotifyValidationResult={[Function]}
  placeholder="PLACEHOLDER"
  prefix="PREFIX"
  resizable={true}
  suffix="SUFFIX"
  underlined={true}
  validateOnFocusIn={false}
  validateOnFocusOut={false}
  validateOnLoad={true}
  value="VALUE"
/>
`;

exports[`<TextField /> with default options matches its snapshot 1`] = `
<TextField
  autoAdjustHeight={false}
  borderless={false}
  className="y-base-text-field y-text-field"
  deferredValidationTime={200}
  errorMessage=""
  multiline={false}
  onBeforeChange={[Function]}
  onChanged={[Function]}
  onGetErrorMessage={[Function]}
  onNotifyValidationResult={[Function]}
  resizable={true}
  underlined={false}
  validateOnFocusIn={false}
  validateOnFocusOut={false}
  validateOnLoad={true}
/>
`;
